### PR TITLE
handle cyclic dependencies in build-chain

### PR DESCRIPTION
**Pull request by gabemontero**
Fixes https://github.com/openshift/origin/issues/3902

@kargakis @deads2k @bparees PTAL

### CODE REVIEW

File: pkg/cmd/cli/describe/chaindescriber.go [hunk 1]
```diff
@@ -92,6 +93,15 @@ func (d *ChainDescriber) Describe(ist *imageapi.ImageStreamTag, includeInputImag
 		return "", NotFoundErr(fmt.Sprintf("%q", ist.Name))
 	}
 
+	markers := buildanalysis.FindCircularBuilds(g, osgraph.DefaultNamer)
+	if len(markers) > 0 {
+		for _, marker := range markers {
+			if strings.Contains(marker.Message, ist.Name) {
+				return fmt.Sprintf("Exiting examination of %s. %s", ist.Name, marker.Message), nil
```
Comment (bparees) [inline pkg/cmd/cli/describe/chaindescriber.go]: i'm not familiar w/ how oc status reports back, but this doesn't look like the right message to go to a user?
also shouldn't this be returning an error?
(and should it be returning at all, or should it be adding this message to a list of messages to be returned along with other analysis output for this IST?)

Comment (gabemontero) [inline pkg/cmd/cli/describe/chaindescriber.go]: On the returning of error, I interpreted the build-chain path as only returning errors when there was a problem in processing data (bad format, etc.).  In this case, the data was correctly analyzed.  It simply found a user error - a misconfiguration.

On the contents of the message, I went back and forth on whether to just return the marker.Message, but thought that somehow repeating the IST they passed in would provide clarity.  But I'm definitely on the fence with that.  Suggestions welcome, but based on @bparees 's feedback I'm minimally thinking of just returning marker.Message, but will wait on additional feedback.

With that, adding the marker.Message to the normal chain-build output (nested is/bc paris, with each is and bc on a separate line, and each line having increased tabbing) definitely did not flow / look valid in my opinion, and is in fact repetitive, as the marker.Message will display the circular chain.

Comment (bparees) [inline pkg/cmd/cli/describe/chaindescriber.go]: can you attach a sample of the new oc status output when a cycle is detected?

Comment (gabemontero) [inline pkg/cmd/cli/describe/chaindescriber.go]: With the code in its current form:
```
Exiting examination of ruby-22-centos7:latest. Cycle detected in build configurations: BuildConfig|example/ruby-22-centos7 -> ImageStreamTag|example/ruby-hello-world:latest -> BuildConfig|example/ruby-hello-world -> ImageStreamTag|example/ruby-something-else:latest -> BuildConfig|example/ruby-something-else -> ImageStreamTag|example/ruby-22-centos7:latest -> BuildConfig|example/ruby-22-centos7
```

If we just went with the `marker.Message`:
```
Cycle detected in build configurations: BuildConfig|example/ruby-22-centos7 -> ImageStreamTag|example/ruby-hello-world:latest -> BuildConfig|example/ruby-hello-world -> ImageStreamTag|example/ruby-something-else:latest -> BuildConfig|example/ruby-something-else -> ImageStreamTag|example/ruby-22-centos7:latest -> BuildConfig|example/ruby-22-centos7
```
When the code did *not* employ the circular build detection, it produced the following when a cyclic dependency existed between 3 build config / image stream tuples:
```
		istag/ruby-22-centos7:latest
			bc/ruby-22-centos7
				istag/ruby-hello-world:latest
					bc/ruby-hello-world
						istag/ruby-something-else:latest
							bc/ruby-something-else
```

Comment (bparees) [inline pkg/cmd/cli/describe/chaindescriber.go]: can you put it in the context of the full oc status output?

Comment (kargakis) [inline pkg/cmd/cli/describe/chaindescriber.go]: > can you put it in the context of the full oc status output?

So you want to display both the chain and the warning? In `oc status` we use it as a warning that users need to ask for explicitly (-v) but in the context of `oadm build-chain` I think it makes sense to print it unconditionally.

Comment (bparees) [inline pkg/cmd/cli/describe/chaindescriber.go]: no i'm just asking you to show me the full output from oc status so i can see where your new message fits in w/ the existing output.

Comment (gabemontero) [inline pkg/cmd/cli/describe/chaindescriber.go]: Using the same yaml, and with the preface that I did not start any builds (not everything in the test config exists), `oc status -v` produces the following (where the cycle message is one of the warnings as @kargakis mentioned):

```
gmontero ~/go/src/github.com/openshift/origin  (issue3902)$ oc status -v
In project example on server https://localhost:8443

bc/ruby-22-centos7 docker build of https://github.com/openshift/ruby-hello-world
  not built yet

bc/ruby-hello-world docker build of https://github.com/openshift/ruby-hello-world
  not built yet

bc/ruby-something-else docker build of https://github.com/openshift/ruby-hello-world
  not built yet

Errors:
  * bc/ruby-22-centos7 is pushing to istag/ruby-hello-world:latest, but the image stream for that tag does not exist.
  * bc/ruby-hello-world is pushing to istag/ruby-something-else:latest, but the image stream for that tag does not exist.
  * bc/ruby-something-else is pushing to istag/ruby-22-centos7:latest, but the image stream for that tag does not exist.
Warnings:
  * Cycle detected in build configurations: bc/ruby-22-centos7 -> istag/ruby-hello-world:latest -> bc/ruby-hello-world -> istag/ruby-something-else:latest -> bc/ruby-something-else -> istag/ruby-22-centos7:latest -> bc/ruby-22-centos7
  * bc/ruby-22-centos7 builds from istag/ruby-22-centos7:latest, but the image stream tag does not exist.
    try: oc start-build bc/ruby-something-else
  * bc/ruby-hello-world builds from istag/ruby-hello-world:latest, but the image stream tag does not exist.
    try: oc start-build bc/ruby-22-centos7
  * bc/ruby-something-else builds from istag/ruby-something-else:latest, but the image stream tag does not exist.
    try: oc start-build bc/ruby-hello-world
  * istag/ruby-hello-world:latest needs to be imported or created by a build.
    try: oc start-build ruby-22-centos7
  * istag/ruby-something-else:latest needs to be imported or created by a build.
    try: oc start-build ruby-hello-world
  * istag/ruby-22-centos7:latest needs to be imported or created by a build.
    try: oc start-build ruby-something-else

View details with 'oc describe <resource>/<name>' or list everything with 'oc get all'.
```

Comment (bparees) [inline pkg/cmd/cli/describe/chaindescriber.go]: thanks, that output looks good to me, i don't think the "Exiting ......" prefix is needed.
I do wonder if Cycle should be lower case.  defer to @smarterclayton

Comment (gabemontero) [inline pkg/cmd/cli/describe/chaindescriber.go]: Cool ... removed prefix, left "Cycle" for now.  Update pushed.

File: pkg/cmd/cli/describe/chaindescriber_test.go [hunk 2]
```diff
@@ -30,6 +30,19 @@ func TestChainDescriber(t *testing.T) {
...
 		{
+			testName:         "circular test",
+			namespaces:       sets.NewString("example"),
+			output:           "",
+			defaultNamespace: "example",
+			name:             "ruby-22-centos7",
+			tag:              "latest",
+			path:             "../../../api/graph/test/circular.yaml",
+			humanReadable: map[string]int{
+				"Cycle detected in build configurations: BuildConfig|example/ruby-22-centos7 -> ImageStreamTag|example/ruby-hello-world:latest -> BuildConfig|example/ruby-hello-world -> ImageStreamTag|example/ruby-something-else:latest -> BuildConfig|example/ruby-something-else -> ImageStreamTag|example/ruby-22-centos7:latest -> BuildConfig|example/ruby-22-centos7": 1,
```
Comment (gabemontero) [inline pkg/cmd/cli/describe/chaindescriber_test.go]: Interesting, the analysis is calling `ResourceName()`:  https://github.com/openshift/origin/blob/master/pkg/build/graph/analysis/bc.go#L148
That produces the `kind|namespace/name` format per https://github.com/openshift/origin/blob/master/pkg/api/graph/types.go#L32

Got to break for a bit, but I'll dig into what produces the `bc/ruby-22-centos7` format and see about switching to that.

Comment (gabemontero) [inline pkg/cmd/cli/describe/chaindescriber_test.go]: I was using the wrong namer .... will update shortly.

Comment (gabemontero) [inline pkg/cmd/cli/describe/chaindescriber_test.go]: update pushed

**Comment by kargakis** [comment]
LGTM [merge]

**Pull request merged**